Option to pass a template to format input

wasiahmad opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Optional input templating for SFT datasets: supply a YAML template to produce a consolidated formatted_input used during preprocessing and message construction for train/validation.

* **Bug Fixes**
  * ASR evaluation now uses unified Whisper-style text normalization for more consistent WER results; ensure samples are marked as ASR where applicable.

* **Chores**
  * Simplified default ASR evaluation/generation flags for clearer behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds optional YAML-driven input templating to combine multiple fields into a formatted_input used for message construction, configurable via data.input_template_path.
> 
> - **Data preprocessing (PromptResponseDataset)**:
>   - Optional templating: load YAML via `load_prompt_config` and render `formatted_input` from multiple fields (semicolon-separated `input_key`). Requires `user` key in template.
>   - Message construction now uses templated `formatted_input` when provided; `add_messages_key` updated to accept `input_key` parameter.
>   - New `apply_input_template` map step applied before message creation; asserts no pre-existing `messages`.
> - **Configuration wiring**:
>   - `data.input_template_path` accepted and passed into dataset setup.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2353efe3aea122a8b27649a8200080f9a473ed14. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

wasiahmad committed:
adding an input template for formatting input

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
adding an input template for formatting input

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
minor bug fix

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
replacing comma with semicolon to avoid hydra issues

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

Kipok commented on nemo_skills/training/nemo_rl/start_sft.py:
```diff
@@ -98,6 +100,14 @@ def __init__(
 
+        self.input_template = None
+        if input_template_path:
+            with open(input_template_path, "rt", encoding="utf-8") as fin:
```
can we instead use nemo-skills functions directly like in grpo example?

Kipok commented on nemo_skills/training/nemo_rl/start_sft.py:
```diff
@@ -86,6 +87,7 @@ def __init__(
         output_key: str = "output",
         num_proc: int | None = None,
         force_reprocess: bool = False,
+        input_template_path: str | None = None,
```
also wouldn't this fail for typical cases when e.g. you have keys like problem and generation in your file, but the logic for setting defaults checks for input / output or messages? So that part should also be updated

wasiahmad replied:
@Kipok I didn't understand this. When a user is passing `input_template_path`, their provided `input_key` field must match with keys present in the `input_template`. Why having keys like problem and generation would cause failure?

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
using nemo-skills load config

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
Merge branch 'main' into apply_input_template

wasiahmad committed:
Merge branch 'main' into apply_input_template

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
resolving conflicts

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
Merge branch 'main' into apply_input_template

Kipok reviewed (changes requested):
added a few more comments. If you want to merge this please also add a new gpu test in https://github.com/NVIDIA-NeMo/Skills/blob/main/tests/gpu-tests/test_train.py to ensure this is being tested in the ci

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
Merge branch 'main' into apply_input_template

melllinia committed:
HF ASR Leaderboard Fix (#1140)

Signed-off-by: mmkrtchyan <mmkrtchyan@nvidia.com>

wasiahmad committed:
Merge remote-tracking branch 'origin/main' into apply_input_template

wasiahmad committed:
Merge branch 'main' into apply_input_template

wasiahmad committed:
Merge branch 'main' into apply_input_template

wasiahmad committed:
Merge branch 'main' into apply_input_template

Made-with: Cursor

wasiahmad committed:
Pass tools from datum into SFT preprocessor formatting

Forward optional tool definitions so message logs match tool-aware templates.

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>
Made-with: Cursor

wasiahmad committed:
SFT debug: print first five messages in message log

Align debug loop with comment when inspecting formatted turns.

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>
Made-with: Cursor

wasiahmad committed:
fixing chat template apply issue with tool calls

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
Merge branch 'main' into apply_input_template

wasiahmad committed:
Align asr-leaderboard dataset __init__ with main

Drop branch-only eval/generation defaults so the PR matches upstream for this file.

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>
Made-with: Cursor

wasiahmad committed:
Fix PromptResponseDataset cache invalidation for SFT

Include data size, input/output keys, and input template file hash in
signature.json so caches rebuild when processing inputs change.

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>
Made-with: Cursor

wasiahmad committed:
adding moe_aux_loss_coeff=0.0 to bypass MoE aux loss validation assertion error

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
adding moe_aux_loss_coeff=0.0 to bypass MoE aux loss validation assertion error

Signed-off-by: wasiahmad <wasiahmad@ucla.edu>

wasiahmad committed:
Merge branch 'main' into apply_input_template
